testCPUID-all/testCPUID.cpp

```diff
@@ -0,0 +1,433 @@
+    size_t waited_count = 0;
+    printf("original value is %Id\n", original_value);
+    uint64_t start = __rdtsc();
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        YieldProcessor();
+        waited_count++;
+    }
+
+    g_waited_count = waited_count;
+
+    uint64_t elapsed = __rdtsc() - start;
+    //printf("[%10s] changed to %Id and waited %Is times, %I64d cycles elapsed!\n", 
+    //    "pause", g_aligned_global_location.loc, formatNumber(waited_count), elapsed);
+
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_tpause(LPVOID lpParam)
+{
+    size_t original_value = g_aligned_global_location.loc;
+    size_t waited_count = 0;
+    printf("original value is %Id, timeout is %I64d\n", original_value, g_timeout);
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        //_umonitor(&g_aligned_global_location.loc);
+
+        //_umwait(1, g_timeout);
+        //_tpause(0, g_timeout);
+        _tpause(1, g_timeout);
+        //printf("wait exited\n");
+        waited_count++;
+    }
+
+    //printf("[%10s] changed to %Id and waited %s times!\n", "tpause", g_aligned_global_location.loc, formatNumber(waited_count));
+    g_waited_count = waited_count;
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_umwait(LPVOID lpParam)
+{
+    size_t original_value = g_aligned_global_location.loc;
+    size_t waited_count = 0;
+    printf("original value is %Id, timeout is %I64d\n", original_value, g_timeout);
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        _umonitor((void*)&g_aligned_global_location.loc);
+        _umwait(1, g_timeout);
+        //printf("wait exited\n");
+        waited_count++;
+    }
+
+    //printf("[%10s] changed to %Id and waited %s times!\n", "umwait", g_aligned_global_location.loc, formatNumber(waited_count));
+
+    g_waited_count = waited_count;
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_mwaitx(LPVOID lpParam)
+{
+    size_t original_value = g_aligned_global_location.loc;
+    size_t waited_count = 0;
+    printf("original value is %Id, timeout is %I64d\n", original_value, g_timeout);
+    uint64_t start_cycles = __rdtsc();
+    auto time_start = std::chrono::steady_clock::now();
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        _mm_monitorx((const void*)&g_aligned_global_location.loc, 0, 0);
+
+        if (g_aligned_global_location.loc == original_value) {
+            waited_count++;
+            _mm_mwaitx(2, 0, (uint32_t)g_timeout);
+        }
+    }
+
+    g_total_cycles_on_thread = __rdtsc() - start_cycles;
+    g_elapsed_time_us_on_thread = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - time_start).count();
+
+    printf("[%10s] changed to %Id and waited %s times!\n", "mwaitx", g_aligned_global_location.loc, formatNumber(waited_count));
+
+    g_waited_count = waited_count;
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_mwaitx2(LPVOID lpParam)
```

Comment:
Is the only difference between `ThreadFunction_mwaitx` and this is when the `waited_count++` (before vs. after `mwaitx`)?

Reply:
right. we can safely get rid of `ThreadFunction_mwaitx2` if we want. I just wanted to try it to see if I can observe any diff